File: 0101-0200/0190.Reverse_Bits.js
//45ms





/**
 * @param {number} n
 * @return {number}
 */
var reverseBits = function(n) {
    let result = 0;
    for (let i = 0; i < 32; i++) {
        result = (result << 1) | (n & 1);
        n >>>= 1;
    }
    return result >>> 0;
};







/**
 * @param {number} n
 * @return {number}
 */
var reverseBits = function(n) {
    let result = 0;               // This will store the reversed bits

    for (let i = 0; i < 32; i++) {  
        // Shift result left to make space for the next bit
        // OR with the last bit of n to append that bit to result
        result = (result << 1) | (n & 1);

        // Unsigned right shift n to process the next bit
        n >>>= 1;
    }

    // Force result to be interpreted as an unsigned 32-bit integer
    return result >>> 0;
};





//40ms







/**
 * @param {number} n
 * @return {number}
 */
var reverseBits = function(n) {
    let res = 0;
    for (let i = 0; i < 32; i++) {
        res = (res << 1) + (n & 1);
        n >>= 1;
    }
    return res;
};







/**
 * @param {number} n
 * @return {number}
 */
var reverseBits = function(n) {
    let res = 0;                         // This will accumulate the reversed bits

    for (let i = 0; i < 32; i++) {       // Loop through all 32 bits
        // Shift res left by 1 to make room for the next bit,
        // then add the least significant bit of n (n & 1)
        res = (res << 1) + (n & 1);

        // Right-shift n by 1 (signed shift),
        // discarding the bit we just used
        n >>= 1;
    }

    // Return the reversed bit sequence
    return res;
};







//48ms







/**
 * @param {number} n
 * @return {number}
 */
var reverseBits = function(n) {
    const binary = n.toString(2).padStart(32,0).split('').reverse().join('');
    return parseInt(binary, 2);
};







/**
 * @param {number} n
 * @return {number}
 */
var reverseBits = function(n) {
    // Convert n to a binary string (base 2),
    // pad it to 32 bits with leading zeros,
    // split into an array of characters,
    // reverse the array,
    // and join it back into a reversed 32-bit binary string.
    const binary = n.toString(2).padStart(32, 0).split('').reverse().join('');

    // Convert the reversed binary string back into a number
    return parseInt(binary, 2);
};

File: 0101-0200/0190.Reverse_Bits.cpp
//4ms





class Solution {
public:
    uint32_t reverseBits(uint32_t n) {
        uint32_t ans=0;
        for (int i = 0; i < 32; i++) {
            ans =  ans<<1;
            if(n&1){
                ans=ans|1;
            }
            n = n>>1;
        }
        return ans;
    }
};




class Solution {
public:
    // Function to reverse the bits of a 32-bit unsigned integer
    uint32_t reverseBits(uint32_t n) {

        uint32_t ans = 0;   // This will store the reversed bit result

        for (int i = 0; i < 32; i++) {   // Process all 32 bits
            ans = ans << 1;              // Shift ans left to make space for the next bit

            if (n & 1) {                 // Check if the least-significant bit of n is 1
                ans = ans | 1;           // If yes, set the last bit of ans to 1
            }

            n = n >> 1;                  // Shift n right to process the next bit
        }

        return ans;      // Final reversed 32-bit number
    }
};






//0ms





class Solution {
public:
    uint32_t reverseBits(uint32_t n) {
        uint32_t ans = 0;

        for(int i = 0; i < 32; i++) {
            ans = ans << 1;      // ans me 1 bit ki jagah bana di
            ans = ans | (n & 1); // n ka last bit ans me daal diya
            n = n >> 1;          // n ka last bit hata diya
        }

        return ans;
    }
};
